File: snake_js/snake.js
const CELL = 20;
const canvas = document.getElementById("game");
const ctx = canvas.getContext("2d");
const COLS = canvas.width / CELL;
const ROWS = canvas.height / CELL;
const scoreEl = document.getElementById("score");

let snake, dir, pending, food, score, alive;

function reset() {
  snake = [{ x: COLS >> 1, y: ROWS >> 1 }, { x: (COLS >> 1) - 1, y: ROWS >> 1 }];
  dir = pending = { x: 1, y: 0 };
  score = 0;
  alive = true;
  scoreEl.textContent = score;
  placeFood();
}

function placeFood() {
  const free = [];
  for (let x = 0; x < COLS; x++)
    for (let y = 0; y < ROWS; y++)
      if (!snake.some(s => s.x === x && s.y === y)) free.push({ x, y });
  food = free.length ? free[Math.floor(Math.random() * free.length)] : null;
}

const KEYS = {
  ArrowUp: [0, -1], w: [0, -1], ArrowDown: [0, 1], s: [0, 1],
  ArrowLeft: [-1, 0], a: [-1, 0], ArrowRight: [1, 0], d: [1, 0],
};

document.addEventListener("keydown", e => {
  if (e.key === "r" || e.key === "R") return reset();
  const k = KEYS[e.key];
  if (!k) return;
  e.preventDefault();
  if (k[0] + dir.x !== 0 || k[1] + dir.y !== 0) pending = { x: k[0], y: k[1] };
});

function step() {
  dir = pending;
  const head = { x: snake[0].x + dir.x, y: snake[0].y + dir.y };
  const eating = food && head.x === food.x && head.y === food.y;
  const body = eating ? snake : snake.slice(0, -1);
  if (head.x < 0 || head.x >= COLS || head.y < 0 || head.y >= ROWS ||
      body.some(s => s.x === head.x && s.y === head.y)) {
    alive = false;
    return;
  }
  snake.unshift(head);
  if (eating) {
    scoreEl.textContent = ++score;
    placeFood();
  } else {
    snake.pop();
  }
}

function draw() {
  ctx.clearRect(0, 0, canvas.width, canvas.height);
  if (food) {
    ctx.fillStyle = "#e33";
    ctx.fillRect(food.x * CELL + 2, food.y * CELL + 2, CELL - 4, CELL - 4);
  }
  snake.forEach((s, i) => {
    ctx.fillStyle = i === 0 ? "#6f6" : "#3a3";
    ctx.fillRect(s.x * CELL + 1, s.y * CELL + 1, CELL - 2, CELL - 2);
  });
  if (!alive) {
    ctx.fillStyle = "#fff";
    ctx.font = "20px sans-serif";
    ctx.textAlign = "center";
    ctx.fillText("Game over - press R to restart", canvas.width / 2, canvas.height / 2);
  }
}

reset();
setInterval(() => { if (alive) step(); draw(); }, 110);

File: snake_python/snake.py
"""Simple Snake game using tkinter. Arrow keys / WASD to move, R to restart."""
import random
import tkinter as tk

CELL = 20
COLS, ROWS = 24, 20
DELAY_MS = 110


class Snake:
    def __init__(self, root):
        self.root = root
        root.title("Snake")
        self.canvas = tk.Canvas(root, width=COLS * CELL, height=ROWS * CELL, bg="#111")
        self.canvas.pack()
        root.bind("<Key>", self.on_key)
        self.reset()
        self.tick()

    def reset(self):
        self.snake = [(COLS // 2, ROWS // 2), (COLS // 2 - 1, ROWS // 2)]
        self.direction = (1, 0)
        self.pending = self.direction
        self.score = 0
        self.alive = True
        self.place_food()

    def place_food(self):
        free = [(x, y) for x in range(COLS) for y in range(ROWS) if (x, y) not in self.snake]
        self.food = random.choice(free) if free else None

    def on_key(self, event):
        keys = {"Up": (0, -1), "w": (0, -1), "Down": (0, 1), "s": (0, 1),
                "Left": (-1, 0), "a": (-1, 0), "Right": (1, 0), "d": (1, 0)}
        if event.keysym in ("r", "R"):
            self.reset()
        elif event.keysym in keys:
            d = keys[event.keysym]
            if (d[0] + self.direction[0], d[1] + self.direction[1]) != (0, 0):
                self.pending = d

    def tick(self):
        if self.alive:
            self.step()
        self.draw()
        self.root.after(DELAY_MS, self.tick)

    def step(self):
        self.direction = self.pending
        x, y = self.snake[0]
        head = (x + self.direction[0], y + self.direction[1])
        eating = head == self.food
        body = self.snake if eating else self.snake[:-1]
        if not (0 <= head[0] < COLS and 0 <= head[1] < ROWS) or head in body:
            self.alive = False
            return
        self.snake.insert(0, head)
        if eating:
            self.score += 1
            self.place_food()
        else:
            self.snake.pop()

    def draw(self):
        c = self.canvas
        c.delete("all")
        if self.food:
            fx, fy = self.food
            c.create_oval(fx * CELL + 2, fy * CELL + 2, (fx + 1) * CELL - 2, (fy + 1) * CELL - 2, fill="#e33")
        for i, (x, y) in enumerate(self.snake):
            c.create_rectangle(x * CELL + 1, y * CELL + 1, (x + 1) * CELL - 1, (y + 1) * CELL - 1,
                               fill="#6f6" if i == 0 else "#3a3", outline="")
        c.create_text(6, 6, anchor="nw", fill="white", text=f"Score: {self.score}")
        if not self.alive:
            c.create_text(COLS * CELL // 2, ROWS * CELL // 2, fill="white", font=("Arial", 16),
                          text="Game over - press R to restart")


if __name__ == "__main__":
    root = tk.Tk()
    Snake(root)
    root.mainloop()
